pkg/job/scraper_test.go

```diff
@@ -0,0 +1,454 @@
+						Roles: []model.Role{
+							{RoleArn: "aws-arn-1", ExternalID: "external-id-1"},
+						},
+					},
+				},
+				CustomNamespaceJobs: []model.CustomNamespaceJob{
+					{
+						Regions:   []string{"us-east-2"},
+						Name:      "my-custom-job",
+						Namespace: "custom-namespace",
+						Roles: []model.Role{
+							{RoleArn: "aws-arn-2", ExternalID: "external-id-2"},
+						},
+					},
+				},
+			},
+			getAccountFunc: func() (string, error) {
+				return "aws-account-1", nil
+			},
+			metadataRunFunc: func(_ context.Context, _ string, _ model.DiscoveryJob) ([]*model.TaggedResource, error) {
+				return []*model.TaggedResource{{
+					ARN: "resource-1", Namespace: "aws-namespace", Region: "us-east-1", Tags: []model.Tag{{Key: "tag1", Value: "value1"}},
+				}}, nil
+			},
+			cloudwatchRunFunc: func(_ context.Context, job cloudwatchrunner.Job) ([]*model.CloudwatchData, error) {
+				if job.Namespace() == "custom-namespace" {
+					return nil, errors.New("I failed you")
+				}
+				return []*model.CloudwatchData{
+					{
+						MetricName:          "metric-1",
+						ResourceName:        "resource-1",
+						Namespace:           "aws-namespace",
+						Tags:                []model.Tag{{Key: "tag1", Value: "value1"}},
+						Dimensions:          []model.Dimension{{Name: "dimension1", Value: "value1"}},
+						GetMetricDataResult: &model.GetMetricDataResult{Statistic: "Maximum", Datapoint: aws.Float64(1.0), Timestamp: time.Time{}},
+					},
+				}, nil
+			},
+			expectedResources: []model.TaggedResourceResult{
+				{
+					Context: &model.ScrapeContext{Region: "us-east-1", AccountID: "aws-account-1"},
+					Data: []*model.TaggedResource{
+						{ARN: "resource-1", Namespace: "aws-namespace", Region: "us-east-1", Tags: []model.Tag{{Key: "tag1", Value: "value1"}}},
+					},
+				},
+			},
+			expectedMetrics: []model.CloudwatchMetricResult{
+				{
+					Context: &model.ScrapeContext{Region: "us-east-1", AccountID: "aws-account-1"},
+					Data: []*model.CloudwatchData{
+						{
+							MetricName:          "metric-1",
+							ResourceName:        "resource-1",
+							Namespace:           "aws-namespace",
+							Tags:                []model.Tag{{Key: "tag1", Value: "value1"}},
+							Dimensions:          []model.Dimension{{Name: "dimension1", Value: "value1"}},
+							GetMetricDataResult: &model.GetMetricDataResult{Statistic: "Maximum", Datapoint: aws.Float64(1.0), Timestamp: time.Time{}},
+						},
+					},
+				},
+			},
+			expectedErrs: []job.Error{
+				{JobContext: job.JobContext{AccountID: "aws-account-1", Namespace: "custom-namespace", Region: "us-east-2", RoleARN: "aws-arn-2"}},
+			},
+		},
+	}
+	for _, tc := range tests {
+		t.Run(tc.name, func(t *testing.T) {
+			rf := testRunnerFactory{
+				GetAccountFunc:    tc.getAccountFunc,
+				MetadataRunFunc:   tc.metadataRunFunc,
+				CloudwatchRunFunc: tc.cloudwatchRunFunc,
+			}
+			sr := job.NewScraper(logging.NewLogger("", true), tc.jobsCfg, &rf)
```

Comment:
(nit) might use the noop logger here.

Reply:
Funny enough, I did this because it was helpful for debugging tests without needing to go to the level of attaching a debugger. I can switch it though it it's too much noise in the test logs.

Reply:
Nope, not a big deal. 👍